### DISCUSSION TITLE
Ambiguous ReportGenerator alias when upgrading to Cake 1.0.0

### DISCUSSION THREAD

**Discussion by Jericho** [cake-build/cake#3196] [General]
This is intended as a heads-up to other devs who might face the same problem I did when upgrading build script to run with Cake 1.0.0. Scroll down for solution.

My build script contains the following task which worked fine up to Cake 0.38.5 but no longer works in Cake 1.0.0

```csharp
Task("Generate-Code-Coverage-Report")
    .IsDependentOn("Run-Code-Coverage")
    .Does(() =>
{
    ReportGenerator(
        "C:\Path\to\my\coverage.xml",
        codeCoverageDir,
        new ReportGeneratorSettings() {
            ClassFilters = new[] { "*.UnitTests*" }
        }
    );
});
```

In fact, my build script does not even build due to the following exception:
```
Error: Error(s) occurred when compiling build script:
D:/_build/ZoomNet/build.cake(232,2): error CS0121: The call is ambiguous between the following methods or properties: 'ReportGenerator(GlobPattern, DirectoryPath, ReportGeneratorSettings)' and 'ReportGenerator(FilePath, DirectoryPath, ReportGeneratorSettings)'
```

This problem appears to be due to the combination of multiple factors.

1. Cake no longer provides a alias for ReportGenerator that accepts a string parameter (as a result of feature #2199) and instead provides a new [alias that accepts a GlobPattern parameter](https://github.com/cake-build/cake/blob/6153bab5989d11bee868d74a7f9b855869ff24bf/src/Cake.Common/Tools/ReportGenerator/ReportGeneratorAliases.cs#L38) in addition to the [alias that accepts a FilePath parameter](https://github.com/cake-build/cake/blob/6153bab5989d11bee868d74a7f9b855869ff24bf/src/Cake.Common/Tools/ReportGenerator/ReportGeneratorAliases.cs#L85) 
2. Both [GlobPattern](https://github.com/cake-build/cake/blob/6153bab5989d11bee868d74a7f9b855869ff24bf/src/Cake.Core/IO/GlobPattern.cs#L38) and [FilePath](https://github.com/cake-build/cake/blob/8bc15841e97944291946a89f335562244965247e/src/Cake.Core/IO/FilePath.cs#L171) define an implicit convertor from string
3. My build script provides a string parameter when invoking the ReportGenerator alias (which incidentally still appears to be the recommendation in the XML documentation [here](https://github.com/cake-build/cake/blob/6153bab5989d11bee868d74a7f9b855869ff24bf/src/Cake.Common/Tools/ReportGenerator/ReportGeneratorAliases.cs#L34) and [here](https://github.com/cake-build/cake/blob/6153bab5989d11bee868d74a7f9b855869ff24bf/src/Cake.Common/Tools/ReportGenerator/ReportGeneratorAliases.cs#L81). I will submit a PR to improve the code sample in XML documentation to avoid this problem).

**Solution**
The solution, turns out to be quite simple. We simply need to specify a `FilePath` parameter when invoking the ReportGenerator alias. Like so:

```csharp
Task("Generate-Code-Coverage-Report")
    .IsDependentOn("Run-Code-Coverage")
    .Does(() =>
{
    ReportGenerator(
        new FilePath("C:\Path\to\my\coverage.xml"),
        codeCoverageDir,
        new ReportGeneratorSettings() {
            ClassFilters = new[] { "*.UnitTests*" }
        }
    );
});
```

**Comment by garywoodfine** [comment]
Thanks this post helped me resolve an issue with was experiencing trying to make use of the Glob.  This post helped me to figure out I needed to use

```csharp
var directory = Directory(TEST_COVERAGE_OUTPUT_DIR);
 var outputDirectory = Directory("./reports");

    var reportSettings = new ReportGeneratorSettings
    {
        ArgumentCustomization = args => args.Append($"-reportTypes:HtmlInline_AzurePipelines")
    };
    
     ReportGenerator(new GlobPattern($"./{ directory}/*.cobertura.xml") , outputDirectory, reportSettings);

```
